tests/test_transform_inline.py

```diff
@@ -281,3 +287,126 @@ def test_constant_replacement_internal(frontend):
     stmts = FindNodes(Assignment).visit(routine.body)
     assert len(stmts) == 1
     assert stmts[0].rhs == 'b + 10'
+
+
+@pytest.mark.parametrize('frontend', available_frontends())
+def test_inline_member_routines(here, frontend):
+    """
+    Test inlining of member subroutines.
+    """
+    fcode = """
+subroutine member_routines(a, b)
+  real(kind=8), intent(inout) :: a(3), b(3)
+  integer :: i
+
+  do i=1, size(a)
+    call add_one(a(i))
+  end do
+
+  call add_to_a(b)
+
+  do i=1, size(a)
+    call add_one(a(i))
+  end do
+
+  contains
+
+    subroutine add_one(a)
+      real(kind=8), intent(inout) :: a
+      a = a + 1
+    end subroutine
+
+    subroutine add_to_a(b)
+      real(kind=8), intent(inout) :: b(:)
+      integer :: n
+
+      n = size(a)
+      do i = 1, n
+        a(i) = a(i) + b(i)
+      end do
+    end subroutine
+end subroutine member_routines
+    """
+    routine = Subroutine.from_source(fcode, frontend=frontend)
+
+    filepath = here/(f'ref_transform_inline_member_routines_{frontend}.f90')
+    reference = jit_compile(routine, filepath=filepath, objname='member_routines')
+
+    a = np.array([1., 2., 3.], order='F')
+    b = np.array([3., 3., 3.], order='F')
+    reference(a, b)
+
+    assert (a == [6., 7., 8.]).all()
+    assert (b == [3., 3., 3.]).all()
+
+    # Now inline the member routines and check again
+    inline_member_procedures(routine=routine)
+
+    assert not routine.members
+    assert not FindNodes(CallStatement).visit(routine.body)
+    assert len(FindNodes(Loop).visit(routine.body)) == 3
+    assert 'n' in routine.variables
+
+    # An verify compiled behaviour
+    filepath = here/(f'transform_inline_member_routines_{frontend}.f90')
+    function = jit_compile(routine, filepath=filepath, objname='member_routines')
+
+    a = np.array([1., 2., 3.], order='F')
+    b = np.array([3., 3., 3.], order='F')
+    function(a, b)
+
+    assert (a == [6., 7., 8.]).all()
+    assert (b == [3., 3., 3.]).all()
+
+
+@pytest.mark.parametrize('frontend', available_frontends())
+def test_inline_member_routines_arg_dimensions(frontend):
+    """
+    Test inlining of member subroutines when sub-arrays of rank less
+    than the formal argument are passed.
+    """
+    fcode = """
+subroutine member_routines_arg_dimensions(matrix, tensor)
+  real(kind=8), intent(inout) :: matrix(3, 3), tensor(3, 3, 4)
+  integer :: i
+  do i=1, 3
+    call add_one(3, matrix(:,i), tensor(1:3,i,:))
```

Comment:
I'm quite impressed that the expansion `b(3,1:n)` -> `tensor(1:3,i,1:n)` works because I don't seem to understand it from the implementation...